torch/nn/parallel/distributed.py

```diff
@@ -253,38 +291,75 @@ def _queue_reduction(self, bucket_idx):
+                        with torch.cuda.device(dev_id):
+                            for reduction_stream in dev_r_streams:
+                                default_stream.wait_stream(reduction_stream)
+
             Variable._execution_engine.queue_callback(sync_reduction_streams)
 
     def _start_reduction_threads(self):
         num_buckets = len(self.bucket_sizes)
-        self._reduction_queues = [queue.Queue() for _ in range(num_buckets)]
+
+        if dist._backend == "nccl":
+            # Let NCCL backend uses a single Queue for all reduction buckets
+            # in order to maintain the NCCL kernel launch order on different
+            # nodes to avoid the deadlock. In other words, we will only maintain
+            # a single reduction thread for this purpose. This is OK since all
+            # distributed NCCL calls are asynchronous
+            self._reduction_queues = [queue.Queue()]
+            self._reduction_streams = [[]]
+        else:
+            self._reduction_queues = [queue.Queue() for _ in range(num_buckets)]
+            self._reduction_streams = [[] for _ in range(num_buckets)]
+
         self._reduction_threads = []
-        self._reduction_streams = [[] for _ in range(num_buckets)]
         self._nccl_streams = []
         self._default_streams = []
+
         for dev_id in self.device_ids:
             with torch.cuda.device(dev_id):
                 # TODO: don't assume we're on a default stream
                 self._default_streams.append(torch.cuda.current_stream())
-                self._nccl_streams.append(torch.cuda.Stream())
+                # NCCL stream is not needed for NCCL backend since current
+                # implmentation uses the default stream
+                if dist._backend != "nccl":
+                    self._nccl_streams.append(torch.cuda.Stream())
+
         for reduction_queue, reduction_streams in zip(self._reduction_queues, self._reduction_streams):
-            for dev_id in self.device_ids:
-                with torch.cuda.device(dev_id):
-                    reduction_streams.append(torch.cuda.Stream())
-            # We only use the first device for distributed reductions
-            dist._register_stream(reduction_streams[0])
+            # Reduction stream not needed for NCCL backend
+            if dist._backend != "nccl":
+                for dev_id in self.device_ids:
+                    with torch.cuda.device(dev_id):
+                        reduction_streams.append(torch.cuda.Stream())
+                # We only use the first device for distributed reductions
+                dist._register_stream(reduction_streams[0])
+
             group_id = dist.new_group()
 
-            self._reduction_threads.append(threading.Thread(
-                target=self._reduction_thread_fn,
-                args=(reduction_queue, group_id, self.device_ids, reduction_streams, self._nccl_streams)))
+            if dist._backend == "nccl":
```

Comment:
you have three conditionals for NCCL in a row. Please concat them to form a single large conditional